imageservice: Add context to various methods

vvoland opened the pull request:
Replaces:
- https://github.com/moby/moby/pull/43828


**- What I did**
I tracked down the integration test hang and reverted the changes which caused it.
I will be doing some further investigations to find out exactly why passing the given context causes this - but for now I propose to keep the existing behaviour of passing `context.TODO` to `container.Run` as this will unblock us in upstreaming the containerd integration changes.

**- How I did it**
Bisecting changes from the original commit

**- How to verify it**
Check if integration tests don't timeout

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->


**- A picture of a cute animal (not mandatory but encouraged)**

ndeloof committed:
imageservice: Add context to various methods

Co-authored-by: Paweł Gronowski <pawel.gronowski@docker.com>
Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland committed:
daemon/start: Revert passing ctx to ctr.Start

This caused integration tests to timeout in the CI

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

corhere commented on daemon/oci_linux.go:
```diff
@@ -1007,7 +1007,7 @@ func WithUser(c *container.Container) coci.SpecOpts {
 }
 
-func (daemon *Daemon) createSpec(c *container.Container) (retSpec *specs.Spec, err error) {
+func (daemon *Daemon) createSpec(ctx context.Context, c *container.Container) (retSpec *specs.Spec, err error) {
```
There's a `context.Background()` call on line 1055 which could be updated.

vvoland replied:
Thanks! I had it in my experimental branch, but forgot to carry it over here :)

corhere commented on daemon/start.go:
```diff
@@ -177,15 +177,17 @@ func (daemon *Daemon) containerStart(container *container.Container, checkpoint
+
 	// TODO(mlaventure): we need to specify checkpoint options here
-	tsk, err := ctr.Start(ctx, checkpointDir,
+	tsk, err := ctr.Start(ctrCtx, checkpointDir,
```
```suggestion
	tsk, err := ctr.Start(context.TODO() /* Passing ctx caused integration tests to be stuck in the cleanup phase */, checkpointDir,
```
Please attach the TODO to the relevant line. It saves readers from having to chase variables around and check that they aren't used in unintended places.

corhere commented on builder/dockerfile/builder.go:
```diff
@@ -244,7 +244,7 @@ func printCommand(out io.Writer, currentCommandIndex int, totalCommands int, cmd
 }
 
-func (b *Builder) dispatchDockerfileWithCancellation(parseResult []instructions.Stage, metaArgs []instructions.ArgCommand, escapeToken rune, source builder.Source) (*dispatchState, error) {
+func (b *Builder) dispatchDockerfileWithCancellation(ctx context.Context, parseResult []instructions.Stage, metaArgs []instructions.ArgCommand, escapeToken rune, source builder.Source) (*dispatchState, error) {
```
`b.clientCtx` is referenced on line 282, below.

corhere commented on builder/dockerfile/builder.go:
```diff
@@ -317,7 +317,7 @@ func (b *Builder) dispatchDockerfileWithCancellation(parseResult []instructions.
 //
 // TODO: Remove?
-func BuildFromConfig(config *container.Config, changes []string, os string) (*container.Config, error) {
+func BuildFromConfig(ctx context.Context, config *container.Config, changes []string, os string) (*container.Config, error) {
```
`context.Background()` on line 330

corhere reviewed (changes requested):
Being a context plumber ain't easy, that's for sure!

vvoland committed:
builder: Don't store context in struct

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

corhere committed:
cmd/dockerd: Rewrite shutdownDaemon to use context timeout

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland commented on daemon/disk_usage.go:
```diff
@@ -14,7 +14,7 @@ import (
 func (daemon *Daemon) ContainerDiskUsage(ctx context.Context) ([]*types.Container, error) {
 	ch := daemon.usage.DoChan("ContainerDiskUsage", func() (interface{}, error) {
 		// Retrieve container list
-		containers, err := daemon.Containers(&types.ContainerListOptions{
+		containers, err := daemon.Containers(context.TODO(), &types.ContainerListOptions{
 			Size: true,
 			All:  true,
 		})
```
Thanks, look good. Do you mind addressing this in a follow up? The behaviour here doesn't change, and I would like to avoid blowing up this PR with a new package being vendored.

corhere replied:
SGTM! 👍

vvoland commented on daemon/daemon.go:
```diff
@@ -526,7 +526,7 @@ func (daemon *Daemon) restore() error {
 				log.WithError(err).Error("failed to prepare mount points for container")
 			}
-			if err := daemon.containerStart(c, "", "", true); err != nil {
+			if err := daemon.containerStart(ctx, c, "", "", true); err != nil {
```
Good catch. There are many other usages of `context.Background()` in this function (mostly semaphore.Acquire though). I think plumbing context without addressing all other usages isn't really useful. Since passing context here is not really crucial for unblocking us, for now I'll just use `context.Background()` here like anything else that accepts context in this function. Hopefully we'll be able to get back to this on future, maybe when refactoring this function completely (it's definitely too big). WDYT?

corhere replied:
[I agree, 100%.](https://github.com/moby/moby/pull/43401#discussion_r842918719)

vvoland commented:
Thanks for a great analysis @corhere, I addressed your comments.
I ripped out the `builder.clientCtx` completely as it didn't seem to be used much and there's no need to hold it in struct.

corhere commented on daemon/start.go:
```diff
@@ -182,13 +182,9 @@ func (daemon *Daemon) containerStart(ctx context.Context, container *container.C
 	// TODO(mlaventure): we need to specify checkpoint options here
-	tsk, err := ctr.Start(ctrCtx, checkpointDir,
-		container.StreamConfig.Stdin() != nil || container.Config.Tty,
+	tsk, err := ctr.Start(context.TODO(), //Passing ctx to ctr.Start caused integration tests to be stuck in the cleanup phase
```
```suggestion
	tsk, err := ctr.Start(context.TODO(), // Passing ctx to ctr.Start caused integration tests to be stuck in the cleanup phase
```
Comments without a leading space are reserved for machine-parseable directives.

vvoland replied:
Thanks!

corhere commented on builder/dockerfile/builder.go:
```diff
@@ -130,24 +129,28 @@ type Builder struct {
 		docker:           options.Backend,
 		idMapping:        options.IDMapping,
-		imageSources:     newImageSources(clientCtx, options),
+		imageSources:     newImageSources(ctx, options),
```
A function call with a context argument but no error return value sets my spidey-senses tingling. Looking at the implementation, it's worse than I feared. On the one hand, it's been like that for five years, but on the other hand it is directly related to plumbing contexts into ImageService. I leave it up to you to decide whether to address it in this PR or in a follow-up.

vvoland replied:
Thanks! Moved the context to `Get` function instead of sneakily capturing it via closure.

corhere reviewed (commented):
Looking great! There's just one place where error-matching needs to be fixed up to handle wrapped errors.

vvoland commented on builder/dockerfile/dispatchers_test.go:
```diff
@@ -462,7 +467,13 @@ func TestRunWithBuildArgs(t *testing.T) {
+	imageProber, err := newImageProber(context.TODO(), mockBackend, nil, false)
+	if err != nil {
+		t.Fatalf("Could not create image prober: %v", err)
+	}
```
Good point!

vvoland commented on daemon/images/cache.go:
```diff
@@ -10,21 +10,25 @@ import (
 		img, err := i.GetImage(ctx, ref, imagetypes.GetImageOpts{})
 		if err != nil {
+			switch err {
+			case context.Canceled, context.DeadlineExceeded:
```
Done, thanks

corhere reviewed (approved):
LGTM, thanks!

vvoland commented:
Thanks! I just squashed the fix-up commits.

cpuguy83 commented on daemon/start.go:
```diff
@@ -177,16 +177,14 @@ func (daemon *Daemon) containerStart(container *container.Container, checkpoint
 	// TODO(mlaventure): we need to specify checkpoint options here
-	tsk, err := ctr.Start(ctx, checkpointDir,
-		container.StreamConfig.Stdin() != nil || container.Config.Tty,
+	tsk, err := ctr.Start(context.TODO(), // Passing ctx to ctr.Start caused integration tests to be stuck in the cleanup phase
```
😞

thaJeztah commented on builder/builder.go:
```diff
@@ -80,7 +80,7 @@ type Result struct {
 type ImageCacheBuilder interface {
 	// MakeImageCache creates a stateful image cache.
-	MakeImageCache(ctx context.Context, cacheFrom []string) ImageCache
+	MakeImageCache(ctx context.Context, cacheFrom []string) (ImageCache, error)
```
Probably would make sense to add this in the first commit as well, as it's fixing up things from that one.

vvoland replied:
Done

thaJeztah commented on builder/dockerfile/builder.go:
```diff
@@ -117,8 +117,7 @@ type Builder struct {
 
-	docker    builder.Backend
-	clientCtx context.Context
+	docker builder.Backend
```
This one wasn't new (added in https://github.com/moby/moby/commit/c44e7a3e632c3ea961cb8c12ba45371f54e6699c), so we can keep it separate.

thaJeztah commented on daemon/daemon.go:
```diff
@@ -1207,7 +1207,7 @@ func (daemon *Daemon) Shutdown(ctx context.Context) error {
 		if ls, err := daemon.Containers(ctx, &types.ContainerListOptions{}); len(ls) != 0 || err != nil {
 			// metrics plugins still need some cleanup
 			daemon.cleanupMetricsPlugins()
-			return nil
+			return err
```
Might make sense to include this one in the first commit as well (this line wasn't new, but the change is needed because of the context changes in that commit)

vvoland replied:
Done

thaJeztah commented on daemon/daemon.go:
```diff
@@ -781,7 +781,7 @@ func NewDaemon(ctx context.Context, config *config.Config, pluginStore *plugin.S
 	defer func() {
 		if err != nil {
-			if err := d.Shutdown(ctx); err != nil {
+			if err := d.Shutdown(context.Background()); err != nil {
```
This looks like a fix-up for the first commit, so would make sense to squash with that one.

Would it be useful to add a comment here, to describe we deliberately use a `context.Background()` here? Or do you think that would be clear from how it's used?

vvoland replied:
Done, added a comment too. Thanks!

thaJeztah commented on daemon/oci_linux.go:
```diff
@@ -1052,7 +1052,7 @@ func (daemon *Daemon) createSpec(ctx context.Context, c *container.Container) (r
 	}
 
-	return &s, coci.ApplyOpts(context.Background(), nil, &containers.Container{
+	return &s, coci.ApplyOpts(ctx, nil, &containers.Container{
```
This looks like a fix-up for the first commit as well

vvoland replied:
Done

thaJeztah commented on builder/dockerfile/builder.go:
```diff
@@ -330,7 +330,7 @@ func BuildFromConfig(ctx context.Context, config *container.Config, changes []st
 		return nil, errdefs.InvalidParameter(err)
 	}
 
-	b, err := newBuilder(context.Background(), builderOptions{
+	b, err := newBuilder(ctx, builderOptions{
```
Same for this one 😅

vvoland replied:
Done

thaJeztah commented on builder/dockerfile/builder.go:
```diff
@@ -148,7 +148,7 @@ func newBuilder(ctx context.Context, options builderOptions) (*Builder, error) {
 		docker:           options.Backend,
 		idMapping:        options.IDMapping,
-		imageSources:     newImageSources(ctx, options),
+		imageSources:     newImageSources(options),
```
This commit looks like a fix-up for the `builder: Don't store context in struct` commit; probably best to squash with that one.

vvoland replied:
Done

thaJeztah reviewed (changes requested):
changes look good; left some comments about commits that should ideally be squashed (let me know if you need help on that)

vvoland commented on builder/dockerfile/dispatchers_test.go:
```diff
@@ -462,7 +462,7 @@ func TestRunWithBuildArgs(t *testing.T) {
 		return imageCache
 	}
-	b.imageProber = newImageProber(nil, mockBackend, nil, false)
+	b.imageProber = newImageProber(context.TODO(), mockBackend, nil, false)
```
Done

vvoland commented on daemon/disk_usage.go:
```diff
@@ -14,7 +14,7 @@ import (
 	ch := daemon.usage.DoChan("ContainerDiskUsage", func() (interface{}, error) {
 		// Retrieve container list
-		containers, err := daemon.Containers(nil, &types.ContainerListOptions{
+		containers, err := daemon.Containers(context.TODO(), &types.ContainerListOptions{
```
Done

vvoland commented on builder/dockerfile/evaluator_test.go:
```diff
@@ -1,7 +1,7 @@
 
 import (
-	ctx "context"
+	"context"
```
Done

vvoland commented:
Thanks @thaJeztah. Squashed most of the commits 😄

thaJeztah reviewed (approved):
LGTM
